src/cpu/x64/jit_uni_reorder.cpp

```diff
@@ -902,6 +904,341 @@ struct jit_uni_reorder_kernel_f32 : public kernel_t, public jit_generator {
+         */
+        ok = true
+                && (utils::one_of(a, /*4,*/ 8, 16)
+                        || utils::one_of(d, /*4,*/ 8, 16))
+                && ((b == 1 && f == 1 && a == e && c == d)
+                        || (c == 1 && e == 1 && a == f && b == d));
+        if (!ok) return false;
+
+        // Do not handle transpose of dimensions other than last 2
+        for (int i = 2; i < p.ndims; ++i) {
+            if (p.nodes[i].is != p.nodes[i].os) {
+                ok = false;
+                break;
+            }
+        }
+
+        return ok;
+    }
+
+    jit_single_blk_kernel(const tr::prb_t &prb)
+        : jit_generator()
+        , prb_(prb)
+        , ker_(nullptr)
+        , ker_tail_(nullptr)
+        , itype_sz(data_type_size(prb_.itype))
+        , otype_sz(data_type_size(prb_.otype))
+        , block_sz(prb.nodes[0].n) {
+        auto input_stride
+                = prb_.nodes[0].is != 1 ? prb_.nodes[0].is : prb_.nodes[1].is;
+        auto output_stride
+                = prb_.nodes[0].os != 1 ? prb_.nodes[0].os : prb_.nodes[1].os;
+
+        if (block_sz == 8) {
+            gen_ker8x8(0, 0, input_stride, output_stride, 8, 8);
+            block_sz = 8;
+        } else if (block_sz == 16) {
+            gen_ker16x16_in_8x8(input_stride, output_stride);
+            block_sz = 16;
+        } else {
+            assert(!"unimplemented");
+        }
+        ret();
+
+        this->ker_ = (void (*)(const void *, void *, const float *))(getCode());
+        if (block_sz == 8) {
+            auto i_tail = input_stride % 8 != 0 ? input_stride % 8 : 8;
+            auto o_tail = output_stride % 8 != 0 ? output_stride % 8 : 8;
+            if (i_tail != o_tail) {
+                // Cases either i_tail != 8 or o_tail != 8
+                this->ker_tail_ = (void (*)(
+                        const void *, void *, const float *))(getCurr());
+                gen_ker8x8(0, 0, input_stride, output_stride, i_tail, o_tail);
+            }
+        } else if (block_sz == 16) {
+            auto i_tail = input_stride % 16 != 0 ? input_stride % 16 : 16;
+            auto o_tail = output_stride % 16 != 0 ? output_stride % 16 : 16;
+            if (i_tail != o_tail) {
+                // Cases either i_tail != 16 or o_tail != 16
+                this->ker_tail_ = (void (*)(
+                        const void *, void *, const float *))(getCurr());
+                gen_ker16x16_in_8x8(
+                        input_stride, output_stride, i_tail, o_tail);
+            } // else we don't need handle tail situation
+        } else {
+            assert(!"unimplemented");
+        }
+        ret();
+
+        this->set_mask_ = (void (*)())(getCurr());
+        auto i_tail = input_stride % 8;
+        auto o_tail = output_stride % 8;
+
+        if (i_tail != 0)
+            gen_setmask(i_tail);
+        else
+            gen_setmask(o_tail);
+        ret();
+
+        this->vzeroupper_ = (void (*)())(getCurr());
```

Comment:
I am not sure this works fine with `use_autogrow` (see [here](https://github.com/oneapi-src/oneDNN/blob/e5bc59afd372b8ff877cb833c67a623c7029692c/src/cpu/x64/jit_generator.hpp#L1417-L1421)).

This also doesn't interplay well with DNNL_JIT_DUMP and VTune/perf integration. Please split into a separate kernels.

Reply:
if `vzeroupper()` really makes sense as a separate function, it is probably worth having it a standalone function, like `cpu::barrier()` (see [here](https://github.com/oneapi-src/oneDNN/blob/master/src/cpu/x64/cpu_barrier.cpp#L102-L105)).

Tagging @kwiersch on his opinion.

Reply:
I think it does not make sense to have 3 separate kernels here unless there is a clear performance advantage. So I would recommend incorporating the set_mask() kernel into the beginning of the main kernel, and likewise calling vzeroupper at the end of the main kernel (in other words, remove the extra ret() calls and move the mask part to the beginning).

Reply:
Hmmm, set mask could go into tail kernel since tail processing is sparse. 2 separate kernels are probably needed because kernel is extremely small, no loop structure or unroll involved, minimum throughput of either 24cycle or 16 load from whatever source (8c case).

Reply:
While I agree with @CaoZhongZ that the kernel is indeed very small, I also definitely agree with @kwiersch on ask to clarify the performance advantage.

I would love to see what it would cost to have the kernel like that:
``` cpp
tr(inp, out, is_tail) {
    test is_tail, is_tail
    jz label_start_tr
    // tail set instructions
label_start_tr:
   // transposition
   vzeroupper
   ret
}
```

1. You will have less `call *addr`
2. You will have one kernel only (that is compatible with `use_autogrow` and VTune / Linux Perf integration

If that hurts performance we will have to think what else we can do. It seems pretty inconvenient (but technically possible) to merge at least `FL` loop into jit kernel to alleviate the overheads.

Reply:
Correct a bit. It was not multiple kernels, it is a single kernel with two parts, normal path and a tail. I put all the branch burden on C++/compiler and left only AVX2 to JIT. It is fine when VTune report the performance data to a single entry and dump JIT get a single file, I have to check the mechanism.

As long as BTB/frontend always get correct path, the performance still hold, which is the case because comparison could be way ahead of branch, and BTB always hit. Set mask could JIT into tail parts, it won't have perceivable performance hit.

So I summary problems that significant:

1. Windows ABI
2. use_autogrow, VTune and DNNL_JIT_DUMP mechanism
3. persistent of mask setting and where to put vzeroupper

I need pondering a while. 🤔️

Reply:
> It was not multiple kernels, it is a single kernel with two parts

For me every call to `getCode()` leads to a new kernel. I agree that (not taking `autogrow` into account) there is only one buffer for generated sequence of kernels. But again mainly the concern here about jit registration, VTune integration, and Xbyak being able to `getCode()` multiple times w/o causing conceptual issues (like [1]).
True single kernel avoids those issue.

The list seems correct to me. IMO, the bullets 2 and 3 should be resolved by using single kernel (it feels I repeat myself somewhere 😅).
I am not sure what do about 1 though. Preserving all registers might be too expensive. Maybe 6 xmm would be enough, or maybe you could try preserving only those that you really need, if you have a tight upper bound. Say, at most you need xmm7, xmm8, you push them. BTW, another dirty hack might be storing xmm{7,8} in, say, upper parts of ymm14, ymm15 that you don't have to preserve. I am not sure, but this might be a little cheaper than pushing them on the stack.

[1] 
``` cpp
void generate() {
  // buffer start = 0x1000
  f1 = getCode(); // 0x1000
  asm instructions ...
  f2 = getCurr(); // 0x1200
  asm instructions ...
  // Oops, buffer has ended, autogrow => reallocate buffer
  // new buffer start 0x2000
  // invalidates f1 and f2
}

// later
call ker->f2(); // SegFault, as f2 == 0x1200, instead of 0x2200
```

Reply:
OK, drop Windows AVX2 support and requires AVX512 for now. I intended to make minimum burden in JIT. If the ABI design was unfriendly for leaf function performance (which should be the most important case), I don't wrestle it too much.